fastafiles = expand("{path}/{files}.fasta", files=config["sortmerna"]["files"], path=DBPATH + "/sortmerna")
fastaindexed = expand("{path}/idx/{files}", files=config["sortmerna"]["files"], path=DBPATH + "/sortmerna")
ref = ':'.join('%s,%s' % (a, b) for a, b in zip(fastafiles, fastaindexed))

rule filter_rna:
    input:
        'Preprocessing/{type}.r1.trimmed.fq',
        'Preprocessing/{type}.r2.trimmed.fq',
        'Preprocessing/{type}.se.trimmed.fq',
        expand(
            "{db}/sortmerna/idx/{files}.{ext}",
            files=config["sortmerna"]["files"],
            ext=['bursttrie_0.dat', 'kmer_0.dat', 'pos_0.dat', 'stats'],
            db=DBPATH)
    output:
        'Preprocessing/{type}.r1.trimmed.rna_filtered.fq',
        'Preprocessing/{type}.r2.trimmed.rna_filtered.fq',
        'Preprocessing/{type}.se.trimmed.rna_filtered.fq',
        'Preprocessing/{type}.r1.trimmed.rna.fq',
        'Preprocessing/{type}.r2.trimmed.rna.fq',
        'Preprocessing/{type}.se.trimmed.rna.fq',
    params:
        ref = ref, mem = str(config["sortmerna"]["mem"])
    shell:
        """
        TMP_R12=$(mktemp --tmpdir={TMPDIR} -t "tmp_XXXXXX.R12")
        {config[sortmerna][scripts_path]}/merge-paired-reads.sh {input[0]} {input[1]} $TMP_R12.fq
        sortmerna --ref {params.ref} --reads $TMP_R12.fq --fastx --aligned $TMP_R12.rrna \
        --other $TMP_R12.rrna_filt -a {THREADS} --paired_in -m {params.mem}
        {config[sortmerna][scripts_path]}/unmerge-paired-reads.sh $TMP_R12.rrna.fq {output[3]} {output[4]}
        {config[sortmerna][scripts_path]}/unmerge-paired-reads.sh $TMP_R12.rrna_filt.fq {output[0]} {output[1]}
        sortmerna --ref {ref} --reads {input[2]} --fastx --aligned {wildcards.step}/{wildcards.type}.se.trimmed.rna \
        --other {wildcards.step}/{wildcards.type}.se.trimmed.rna_filtered -a {THREADS} -m {params.mem}
        rm $TMP_R12*
        """
